<a href="https://colab.research.google.com/github/amitoshacharya/Movie-Recommender-System/blob/main/Movie_Recommender_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***About Dataset***

    Context

The files contain metadata for all 45,000 movies listed in the Full MovieLens Dataset. The dataset consists of movies released on or before July 2017. Data points include cast, crew, plot keywords, budget, revenue, posters, release dates, languages, production companies, countries, TMDB vote counts and vote averages.

This dataset also has files containing 26 million ratings from 270,000 users for all 45,000 movies. Ratings are on a scale of 1-5 and have been obtained from the official GroupLens website.

    Content
This dataset consists of the following files:

***movies_metadata.csv:***

*The main Movies Metadata file. Contains information on 45,000 movies featured in the Full MovieLens dataset. Features include posters, backdrops, budget, revenue, release dates, languages, production countries and companies.*

***keywords.csv:***

*Contains the movie plot keywords for our MovieLens movies. Available in the form of a stringified JSON Object.*

***credits.csv:*** 

*Consists of Cast and Crew Information for all our movies. Available in the form of a stringified JSON Object.*

***links.csv:*** 

*The file that contains the TMDB and IMDB IDs of all the movies featured in the Full MovieLens dataset.*

***links_small.csv:***

*Contains the TMDB and IMDB IDs of a small subset of 9,000 movies of the Full Dataset.*

***ratings_small.csv:*** 

*The subset of 100,000 ratings from 700 users on 9,000 movies.*

The Full MovieLens Dataset consisting of 26 million ratings and 750,000 tag applications from 270,000 users on all the 45,000 movies in this dataset can be accessed here

    Acknowledgements
This dataset is an ensemble of data collected from TMDB and GroupLens.
The Movie Details, Credits and Keywords have been collected from the TMDB Open API. This product uses the TMDb API but is not endorsed or certified by TMDb. Their API also provides access to data on many additional movies, actors and actresses, crew members, and TV shows.

The Movie Links and Ratings have been obtained from the Official GroupLens website. The files are a part of the dataset available here



    Inspiration
This dataset was assembled as part of my second Capstone Project for Springboard's Data Science Career Track. I wanted to perform an extensive EDA on Movie Data to narrate the history and the story of Cinema and use this metadata in combination with MovieLens ratings to build various types of Recommender Systems.



***Some of the things you can do with this dataset:***

Predicting movie revenue and/or movie success based on a certain metric. What movies tend to get higher vote counts and vote averages on TMDB? Building Content Based and Collaborative Filtering Based Recommendation Engines.

In [ ]:
# importing basic packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

####  preprocessing 1  ####
#read csv movies_metadata
df1= pd.read_csv(r'/content/drive/MyDrive/DATA SCIENCE/Netflix Analysis/movies_metadata.csv')
df1.head()

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [ ]:
df1.shape #row count

(45466, 24)

In [ ]:
df1[:1].T  #Transposing for an idea about 1st records

0
adult                                                              False
belongs_to_collection  {'id': 10194, 'name': 'Toy Story Collection', ...
budget                                                          30000000
genres                 [{'id': 16, 'name': 'Animation'}, {'id': 35, '...
homepage                            http://toystory.disney.com/toy-story
id                                                                   862
imdb_id                                                        tt0114709
original_language                                                     en
original_title                                                 Toy Story
overview               Led by Woody, Andy's toys live happily in his ...
popularity                                                     21.946943
poster_path                             /rhIRbceoE9lR4veEXuwCC2wARtG.jpg
production_companies      [{'name': 'Pixar Animation Studios', 'id': 3}]
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...
release_date                                                  1995-10-30
revenue                                                      373554033.0
runtime                                                             81.0
spoken_languages                [{'iso_639_1': 'en', 'name': 'English'}]
status                                                          Released
tagline                                                              NaN
title                                                          Toy Story
video                                                              False
vote_average                                                         7.7
vote_count                                                        5415.0

In [ ]:
#reorganizing the table columns and eliminating 'original_title'
df2=df1[['id','imdb_id','title','original_language','spoken_languages','tagline','homepage','genres','overview','belongs_to_collection','adult','budget','status','release_date','revenue','runtime','popularity','poster_path','production_companies','production_countries','video','vote_average','vote_count']]
df2.head(3)

id    imdb_id             title original_language  \
0    862  tt0114709         Toy Story                en   
1   8844  tt0113497           Jumanji                en   
2  15602  tt0113228  Grumpier Old Men                en   

                                    spoken_languages  \
0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2           [{'iso_639_1': 'en', 'name': 'English'}]   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   

                               homepage  \
0  http://toystory.disney.com/toy-story   
1                                   NaN   
2                                   NaN   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   

                               belongs_to_collection  ... release_date  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...   1995-10-30   
1                                                NaN  ...   1995-12-15   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...  ...   1995-12-22   

       revenue runtime popularity                       poster_path  \
0  373554033.0    81.0  21.946943  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  262797249.0   104.0  17.015539  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2          0.0   101.0    11.7129  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   

                                production_countries  video vote_average  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...  False          7.7   
1  [{'iso_3166_1': 'US', 'name': 'United States o...  False          6.9   
2  [{'iso_3166_1': 'US', 'name': 'United States o...  False          6.5   

  vote_count  
0     5415.0  
1     2413.0  
2       92.0  

[3 rows x 23 columns]

In [ ]:
df2[:1].T  #Transposing for an idea about 1st records

0
id                                                                   862
imdb_id                                                        tt0114709
title                                                          Toy Story
original_language                                                     en
spoken_languages                [{'iso_639_1': 'en', 'name': 'English'}]
tagline                                                              NaN
homepage                            http://toystory.disney.com/toy-story
genres                 [{'id': 16, 'name': 'Animation'}, {'id': 35, '...
overview               Led by Woody, Andy's toys live happily in his ...
belongs_to_collection  {'id': 10194, 'name': 'Toy Story Collection', ...
adult                                                              False
budget                                                          30000000
status                                                          Released
release_date                                                  1995-10-30
revenue                                                      373554033.0
runtime                                                             81.0
popularity                                                     21.946943
poster_path                             /rhIRbceoE9lR4veEXuwCC2wARtG.jpg
production_companies      [{'name': 'Pixar Animation Studios', 'id': 3}]
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...
video                                                              False
vote_average                                                         7.7
vote_count                                                        5415.0

In [ ]:
#cleaning, converting the dict. to list
df2.count()

id                       45466
imdb_id                  45449
title                    45460
original_language        45455
spoken_languages         45460
tagline                  20412
homepage                  7782
genres                   45466
overview                 44512
belongs_to_collection     4494
adult                    45466
budget                   45466
status                   45379
release_date             45379
revenue                  45460
runtime                  45203
popularity               45461
poster_path              45080
production_companies     45463
production_countries     45463
video                    45460
vote_average             45460
vote_count               45460
dtype: int64

In [ ]:
#total records
df2.shape 

(45466, 23)

In [ ]:
from ast import literal_eval #https://www.educative.io/answers/what-is-astliteralevalnodeorstring-in-python
df2.genres=df2['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x]) #converting string to python list of genres 

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
df2[:2]

id    imdb_id      title original_language  \
0   862  tt0114709  Toy Story                en   
1  8844  tt0113497    Jumanji                en   

                                    spoken_languages  \
0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   

                                     tagline  \
0                                        NaN   
1  Roll the dice and unleash the excitement!   

                               homepage                        genres  \
0  http://toystory.disney.com/toy-story   [Animation, Comedy, Family]   
1                                   NaN  [Adventure, Fantasy, Family]   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   

                               belongs_to_collection  ... release_date  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...   1995-10-30   
1                                                NaN  ...   1995-12-15   

       revenue runtime popularity                       poster_path  \
0  373554033.0    81.0  21.946943  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  262797249.0   104.0  17.015539  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   

                                production_countries  video vote_average  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...  False          7.7   
1  [{'iso_3166_1': 'US', 'name': 'United States o...  False          6.9   

  vote_count  
0     5415.0  
1     2413.0  

[2 rows x 23 columns]

 # ***Rating for Top Movies filtering***

    Weighted Rating (WR) =  [v/(v+m)] X R + [m/(v+m)] X C 

where,

v is the number of votes for the movie

m is the minimum votes required to be listed in the chart

R is the average rating of the movie

C is the mean vote across the whole report

The next step is to determine an appropriate value for ***m, the minimum votes required to be listed in the chart***. 

We will use 95th percentile as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 95% of the movies in the list.

In [ ]:
df3= df2[['title', 'genres', 'release_date', 'popularity', 'vote_average', 'vote_count']]
df3.head()

title                        genres release_date  \
0                    Toy Story   [Animation, Comedy, Family]   1995-10-30   
1                      Jumanji  [Adventure, Fantasy, Family]   1995-12-15   
2             Grumpier Old Men             [Romance, Comedy]   1995-12-22   
3            Waiting to Exhale      [Comedy, Drama, Romance]   1995-12-22   
4  Father of the Bride Part II                      [Comedy]   1995-02-10   

  popularity  vote_average  vote_count  
0  21.946943           7.7      5415.0  
1  17.015539           6.9      2413.0  
2    11.7129           6.5        92.0  
3   3.859495           6.1        34.0  
4   8.387519           5.7       173.0

In [ ]:
df3['year']= pd.to_datetime(df3.release_date, errors='coerce').apply(lambda x: str(x).split('-')[0])
df3.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


title                        genres release_date  \
0                    Toy Story   [Animation, Comedy, Family]   1995-10-30   
1                      Jumanji  [Adventure, Fantasy, Family]   1995-12-15   
2             Grumpier Old Men             [Romance, Comedy]   1995-12-22   
3            Waiting to Exhale      [Comedy, Drama, Romance]   1995-12-22   
4  Father of the Bride Part II                      [Comedy]   1995-02-10   

  popularity  vote_average  vote_count  year  
0  21.946943           7.7      5415.0  1995  
1  17.015539           6.9      2413.0  1995  
2    11.7129           6.5        92.0  1995  
3   3.859495           6.1        34.0  1995  
4   8.387519           5.7       173.0  1995

In [ ]:
df3.vote_count=df3[df3.vote_count.notnull()].vote_count.astype('int')
df3.vote_average=df3[df3.vote_average.notnull()].vote_average.astype('int')

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
#dataset level parameter
m= df3['vote_count'].quantile(0.95) #m is the minimum votes required to be listed in the chart. Here we are considering 95% of 
C= df3['vote_average'].mean() #C is the mean vote across the whole report

m,C

(434.0, 5.244896612406511)

In [ ]:
#Qualify for Top movies
qualified_movies= df3[(df3.vote_count >= m) & (df3.vote_count.notnull()) & df3.vote_average.notnull()].drop('release_date', axis=1)
qualified_movies.shape

(2274, 6)

###***Conclusion***
Therefore, to qualify to be considered for the chart, a movie has to have at least 434 votes counts. 

We also see that the average rating for a movie is 5.244 on a scale of 10. 

Thus, 2274 Movies qualify to be on our chart.

In [ ]:
def wr(dataframe):
  v= dataframe['vote_count'] #v is the number of votes for the movie
  R= dataframe['vote_average'] #R is the average rating of the movie
  
  return ((v/(v+m)) * R) + (m/(v+m) * C)


In [ ]:
qualified_movies['wr'] = qualified_movies.apply(wr, axis=1)
qualified_movies= qualified_movies.sort_values('wr', ascending=False).head(500)

#***TOP MOVIES Recommendations***

New user will be populated with the top movies list

In [ ]:
qualified_movies.head(20)

title  \
15480                                          Inception   
12481                                    The Dark Knight   
22879                                       Interstellar   
2843                                          Fight Club   
4863   The Lord of the Rings: The Fellowship of the Ring   
292                                         Pulp Fiction   
314                             The Shawshank Redemption   
7000       The Lord of the Rings: The Return of the King   
351                                         Forrest Gump   
5814               The Lord of the Rings: The Two Towers   
256                                            Star Wars   
1225                                  Back to the Future   
834                                        The Godfather   
1154                             The Empire Strikes Back   
46                                                 Se7en   
24860                                 The Imitation Game   
359                                        The Lion King   
18465                                   The Intouchables   
22841                           The Grand Budapest Hotel   
586                             The Silence of the Lambs   

                                                  genres  popularity  \
15480  [Action, Thriller, Science Fiction, Mystery, A...   29.108149   
12481                   [Drama, Action, Crime, Thriller]  123.167259   
22879                [Adventure, Drama, Science Fiction]   32.213481   
2843                                             [Drama]   63.869599   
4863                        [Adventure, Fantasy, Action]   32.070725   
292                                    [Thriller, Crime]  140.950236   
314                                       [Drama, Crime]   51.645403   
7000                        [Adventure, Fantasy, Action]   29.324358   
351                             [Comedy, Drama, Romance]   48.307194   
5814                        [Adventure, Fantasy, Action]   29.423537   
256                 [Adventure, Action, Science Fiction]   42.149697   
1225        [Adventure, Comedy, Science Fiction, Family]   25.778509   
834                                       [Drama, Crime]   41.109264   
1154                [Adventure, Action, Science Fiction]   19.470959   
46                            [Crime, Mystery, Thriller]    18.45743   
24860                    [History, Drama, Thriller, War]    31.59594   
359                           [Family, Animation, Drama]   21.605761   
18465                                    [Drama, Comedy]   16.086919   
22841                                    [Comedy, Drama]   14.442048   
586                             [Crime, Drama, Thriller]    4.307222   

       vote_average  vote_count  year        wr  
15480           8.0     14075.0  2010  7.917588  
12481           8.0     12269.0  2008  7.905871  
22879           8.0     11187.0  2014  7.897107  
2843            8.0      9678.0  1999  7.881753  
4863            8.0      8892.0  2001  7.871787  
292             8.0      8670.0  1994  7.868660  
314             8.0      8358.0  1994  7.864000  
7000            8.0      8226.0  2003  7.861927  
351             8.0      8147.0  1994  7.860656  
5814            8.0      7641.0  2002  7.851924  
256             8.0      6778.0  1977  7.834205  
1225            8.0      6239.0  1985  7.820813  
834             8.0      6024.0  1972  7.814847  
1154            8.0      5998.0  1980  7.814099  
46              8.0      5915.0  1995  7.811669  
24860           8.0      5895.0  2014  7.811074  
359             8.0      5520.0  1994  7.799175  
18465           8.0      5410.0  2011  7.795394  
22841           8.0      4644.0  2014  7.764530  
586             8.0      4549.0  1991  7.760041

In [ ]:
s = df3.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True) #segregating each movies with individual genres records
s.name= 'genres'
genres_df = df3.drop(['genres','release_date'], axis=1).join(s)
genres_df.head(10)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  """Entry point for launching an IPython kernel.


title popularity  vote_average  vote_count  year     genres
0          Toy Story  21.946943           7.0      5415.0  1995  Animation
0          Toy Story  21.946943           7.0      5415.0  1995     Comedy
0          Toy Story  21.946943           7.0      5415.0  1995     Family
1            Jumanji  17.015539           6.0      2413.0  1995  Adventure
1            Jumanji  17.015539           6.0      2413.0  1995    Fantasy
1            Jumanji  17.015539           6.0      2413.0  1995     Family
2   Grumpier Old Men    11.7129           6.0        92.0  1995    Romance
2   Grumpier Old Men    11.7129           6.0        92.0  1995     Comedy
3  Waiting to Exhale   3.859495           6.0        34.0  1995     Comedy
3  Waiting to Exhale   3.859495           6.0        34.0  1995      Drama

##***Top Genre Based Recommender***

In [ ]:
#top movies based on genres 
#for now we will use 85% as our cut-off value.
def top_movies(genre, hd, percentile=0.85):
  gen= genres_df[genres_df.genres == genre].copy()

  m= gen.vote_count.quantile(percentile)
  C= gen.vote_average.mean()

  qualified = gen[(gen.vote_count >= m) & gen.vote_count.notnull() & gen.vote_average.notnull()]

  v= qualified.vote_count
  R= qualified.vote_average

  wr= ((v/(v+m)) * R) + (m/(v+m) * C)

  qualified['wr'] = wr
  return qualified.sort_values('wr', ascending=False).head(hd)

In [ ]:
#top 10 Romance movies
top_movies(genre='Romance', hd=10)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  app.launch_new_instance()


title popularity  vote_average  vote_count  year  \
10309  Dilwale Dulhania Le Jayenge  34.457024           9.0       661.0  1995   
351                   Forrest Gump  48.307194           8.0      8147.0  1994   
876                        Vertigo   18.20822           8.0      1162.0  1958   
40251                   Your Name.  34.461252           8.0      1030.0  2016   
883               Some Like It Hot  11.845107           8.0       835.0  1959   
1132               Cinema Paradiso  14.177005           8.0       834.0  1988   
19901                     Paperman   7.198633           8.0       734.0  2012   
37863                  Sing Street  10.672862           8.0       669.0  2016   
882                  The Apartment  11.994281           8.0       498.0  1960   
38718               The Handmaiden  16.727405           8.0       453.0  2016   

        genres        wr  
10309  Romance  8.565285  
351    Romance  7.971357  
876    Romance  7.811667  
40251  Romance  7.789489  
883    Romance  7.745154  
1132   Romance  7.744878  
19901  Romance  7.713951  
37863  Romance  7.689483  
882    Romance  7.599317  
38718  Romance  7.566166

In [ ]:
#top 10 Adventure movies
top_movies(genre='Adventure', hd=10)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  app.launch_new_instance()


title popularity  \
15480                                          Inception  29.108149   
22879                                       Interstellar  32.213481   
4863   The Lord of the Rings: The Fellowship of the Ring  32.070725   
7000       The Lord of the Rings: The Return of the King  29.324358   
5814               The Lord of the Rings: The Two Towers  29.423537   
256                                            Star Wars  42.149697   
1225                                  Back to the Future  25.778509   
1154                             The Empire Strikes Back  19.470959   
5481                                       Spirited Away  41.048867   
9698                                Howl's Moving Castle  16.136048   

       vote_average  vote_count  year     genres        wr  
15480           8.0     14075.0  2010  Adventure  7.906526  
22879           8.0     11187.0  2014  Adventure  7.883426  
4863            8.0      8892.0  2001  Adventure  7.854939  
7000            8.0      8226.0  2003  Adventure  7.843867  
5814            8.0      7641.0  2002  Adventure  7.832647  
256             8.0      6778.0  1977  Adventure  7.812801  
1225            8.0      6239.0  1985  Adventure  7.797828  
1154            8.0      5998.0  1980  Adventure  7.790329  
5481            8.0      3968.0  2001  Adventure  7.695056  
9698            8.0      2049.0  2004  Adventure  7.465435

####***Conclusion***
According to matrix the Top Romance movie is Bollywood's ***'Dilwale Dulhania Le Jayenge'*** whereas Top Adventure movie is Hollywood's ***'Inception'***

#***Content Based Recommendor (CBR)***

The recommender in the previous section suffers some severe limitations. For one, it gives the same recommendation to everyone, regardless of the user's personal taste. If a person who loves romantic movies (and hates action) were to look at our Top 15 Chart, s/he wouldn't probably like most of the movies. If s/he were to go one step further and look at our charts by genre, s/he wouldn't still be getting the best recommendations.

For instance, consider a person who loves Dilwale Dulhania Le Jayenge, My Name is Khan and Kabhi Khushi Kabhi Gham. One inference we can obtain is that the person loves the actor Shahrukh Khan and the director Karan Johar. Even if s/he were to access the romance chart, s/he wouldn't find these as the top recommendations.

To personalise the recommendations more, we need to build an engine that computes similarity between movies based on certain metrics and suggests movies that are most similar to a particular movie that a user liked. Since we will be using movie metadata (or content) to build this engine, this also known as ***Content Based Filtering***.

I will build two Content Based Recommenders based on:

*   Movie Overviews and Taglines
*   Movie Cast, Crew, Keywords and Genre



In [ ]:
links_sml=pd.read_csv(r'/content/drive/MyDrive/DATA SCIENCE/Netflix Analysis/links_small.csv')
links_sml.head()

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0

In [ ]:
links_sml=links_sml[links_sml.tmdbId.notnull()]['tmdbId'].astype('int')
links_sml.head()

0      862
1     8844
2    15602
3    31357
4    11862
Name: tmdbId, dtype: int64

In [ ]:
#EDA of metadata as md
md= df2.copy()
md=md.drop([19730, 29503, 35587])
md.id=md.id.astype('int')
md.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45463 entries, 0 to 45465
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     45463 non-null  int64  
 1   imdb_id                45446 non-null  object 
 2   title                  45460 non-null  object 
 3   original_language      45452 non-null  object 
 4   spoken_languages       45460 non-null  object 
 5   tagline                20412 non-null  object 
 6   homepage               7779 non-null   object 
 7   genres                 45463 non-null  object 
 8   overview               44509 non-null  object 
 9   belongs_to_collection  4491 non-null   object 
 10  adult                  45463 non-null  object 
 11  budget                 45463 non-null  object 
 12  status                 45379 non-null  object 
 13  release_date           45376 non-null  object 
 14  revenue                45460 non-null  float64
 15  ru

In [ ]:
md.head()

id    imdb_id                        title original_language  \
0    862  tt0114709                    Toy Story                en   
1   8844  tt0113497                      Jumanji                en   
2  15602  tt0113228             Grumpier Old Men                en   
3  31357  tt0114885            Waiting to Exhale                en   
4  11862  tt0113041  Father of the Bride Part II                en   

                                    spoken_languages  \
0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2           [{'iso_639_1': 'en', 'name': 'English'}]   
3           [{'iso_639_1': 'en', 'name': 'English'}]   
4           [{'iso_639_1': 'en', 'name': 'English'}]   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
3  Friends are the people who let you be yourself...   
4  Just When His World Is Back To Normal... He's ...   

                               homepage                        genres  \
0  http://toystory.disney.com/toy-story   [Animation, Comedy, Family]   
1                                   NaN  [Adventure, Fantasy, Family]   
2                                   NaN             [Romance, Comedy]   
3                                   NaN      [Comedy, Drama, Romance]   
4                                   NaN                      [Comedy]   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                               belongs_to_collection  ... release_date  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...   1995-10-30   
1                                                NaN  ...   1995-12-15   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...  ...   1995-12-22   
3                                                NaN  ...   1995-12-22   
4  {'id': 96871, 'name': 'Father of the Bride Col...  ...   1995-02-10   

       revenue runtime popularity                       poster_path  \
0  373554033.0    81.0  21.946943  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  262797249.0   104.0  17.015539  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2          0.0   101.0    11.7129  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   
3   81452156.0   127.0   3.859495  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg   
4   76578911.0   106.0   8.387519  /e64sOI48hQXyru7naBFyssKFxVd.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries  video vote_average  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...  False          7.7   
1  [{'iso_3166_1': 'US', 'name': 'United States o...  False          6.9   
2  [{'iso_3166_1': 'US', 'name': 'United States o...  False          6.5   
3  [{'iso_3166_1': 'US', 'name': 'United States o...  False          6.1   
4  [{'iso_3166_1': 'US', 'name': 'United States o...  False          5.7   

  vote_count  
0     5415.0  
1     2413.0  
2       92.0  
3       34.0  
4      173.0  

[5 rows x 23 columns]

In [ ]:
#filtering small movies from our original metadata dataset
smd=md[md.id.isin(links_sml)]
smd.shape

(9099, 23)

###***Conclusion***

Therefore, 9099 movies avaiable in our small movies metadata dataset which is 5 times smaller than our original dataset of 45000 movies.

In [ ]:
#using small movies dataset for our further CBR
smd.head()

id    imdb_id                        title original_language  \
0    862  tt0114709                    Toy Story                en   
1   8844  tt0113497                      Jumanji                en   
2  15602  tt0113228             Grumpier Old Men                en   
3  31357  tt0114885            Waiting to Exhale                en   
4  11862  tt0113041  Father of the Bride Part II                en   

                                    spoken_languages  \
0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2           [{'iso_639_1': 'en', 'name': 'English'}]   
3           [{'iso_639_1': 'en', 'name': 'English'}]   
4           [{'iso_639_1': 'en', 'name': 'English'}]   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
3  Friends are the people who let you be yourself...   
4  Just When His World Is Back To Normal... He's ...   

                               homepage                        genres  \
0  http://toystory.disney.com/toy-story   [Animation, Comedy, Family]   
1                                   NaN  [Adventure, Fantasy, Family]   
2                                   NaN             [Romance, Comedy]   
3                                   NaN      [Comedy, Drama, Romance]   
4                                   NaN                      [Comedy]   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                               belongs_to_collection  ... release_date  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...   1995-10-30   
1                                                NaN  ...   1995-12-15   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...  ...   1995-12-22   
3                                                NaN  ...   1995-12-22   
4  {'id': 96871, 'name': 'Father of the Bride Col...  ...   1995-02-10   

       revenue runtime popularity                       poster_path  \
0  373554033.0    81.0  21.946943  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  262797249.0   104.0  17.015539  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2          0.0   101.0    11.7129  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   
3   81452156.0   127.0   3.859495  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg   
4   76578911.0   106.0   8.387519  /e64sOI48hQXyru7naBFyssKFxVd.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries  video vote_average  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...  False          7.7   
1  [{'iso_3166_1': 'US', 'name': 'United States o...  False          6.9   
2  [{'iso_3166_1': 'US', 'name': 'United States o...  False          6.5   
3  [{'iso_3166_1': 'US', 'name': 'United States o...  False          6.1   
4  [{'iso_3166_1': 'US', 'name': 'United States o...  False          5.7   

  vote_count  
0     5415.0  
1     2413.0  
2       92.0  
3       34.0  
4      173.0  

[5 rows x 23 columns]

##***`Movie Description Based Recommendor`***

In [ ]:
#create combinations of classification features- tagline and overview
smd['tagline'],smd['overview']= smd.tagline.fillna(''),smd.overview.fillna('')
smd['description']= smd.tagline + smd.overview
smd.decription=smd.description.fillna('')

smd.description

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#a

0        Led by Woody, Andy's toys live happily in his ...
1        Roll the dice and unleash the excitement!When ...
2        Still Yelling. Still Fighting. Still Ready for...
3        Friends are the people who let you be yourself...
4        Just When His World Is Back To Normal... He's ...
                               ...                        
40224    A god incarnate. A city doomed.From the mind b...
40503    The band you know. The story you don't.The ban...
44821    Pokémon: Spell of the UnknownWhen Molly Hale's...
44826    All your favorite Pokémon characters are back,...
45265    While holidaying in the French Alps, a Swedish...
Name: description, Length: 9099, dtype: object

In [ ]:
#vectors for CBR
from sklearn.feature_extraction.text import TfidfVectorizer as vctrizer
tfidfvectors= vctrizer(stop_words='english', ngram_range=(1, 2))
vectors= tfidfvectors.fit_transform(smd.description)

In [ ]:
tfidfvectors.get_feature_names_out()

array(['00', '00 agent', '00 body', ..., 'ˈfil noun', 'ˌrän', 'ˌrän tō'],
      dtype=object)

In [ ]:
#find the similarity b/w vectors using cosine similarity
from sklearn.metrics.pairwise import cosine_similarity
cos_sim= cosine_similarity(vectors, vectors)
cos_sim_df= pd.DataFrame(data=cos_sim)
cos_sim_df.head()

0         1         2     3         4         5         6         7     \
0  1.000000  0.006802  0.000000   0.0  0.000000  0.000000  0.000000  0.000000   
1  0.006802  1.000000  0.015379   0.0  0.002249  0.014953  0.000000  0.000000   
2  0.000000  0.015379  1.000000   0.0  0.007934  0.000000  0.000000  0.002356   
3  0.000000  0.000000  0.000000   1.0  0.000000  0.002174  0.000000  0.006529   
4  0.000000  0.002249  0.007934   0.0  1.000000  0.000000  0.008296  0.000000   

       8     9     ...      9089      9090      9091      9092      9093  \
0  0.000000   0.0  ...  0.000000  0.005668  0.000000  0.000000  0.000000   
1  0.033264   0.0  ...  0.000000  0.006688  0.000000  0.000000  0.000000   
2  0.000000   0.0  ...  0.000000  0.000000  0.000000  0.002318  0.001945   
3  0.000000   0.0  ...  0.016824  0.000000  0.003569  0.001991  0.000000   
4  0.011431   0.0  ...  0.000000  0.000000  0.006035  0.003366  0.000000   

       9094      9095      9096      9097      9098  
0  0.000000  0.004794  0.000000  0.003448  0.000000  
1  0.001754  0.003675  0.003568  0.007623  0.000000  
2  0.002223  0.000000  0.000000  0.002883  0.004737  
3  0.000000  0.000000  0.000000  0.008676  0.000000  
4  0.003822  0.001879  0.004378  0.001395  0.000000  

[5 rows x 9099 columns]

***NOTE:***

***Here in the matrix, we are checking similarity b/w one description as index and others as columns***

In [ ]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title']) #find index from title
indices[:10]

title
Toy Story                      0
Jumanji                        1
Grumpier Old Men               2
Waiting to Exhale              3
Father of the Bride Part II    4
Heat                           5
Sabrina                        6
Tom and Huck                   7
Sudden Death                   8
GoldenEye                      9
dtype: int64

In [ ]:
def cbr_recommendor(title, top_n=5):
  idx= indices[title]

  sim_scores= list(enumerate(cos_sim_df[idx]))
  sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
  sim_scores = sim_scores[1:top_n+1] #removing the first one

  movie_sim_idx=[tup[0] for tup in sim_scores]
  movie_sim=[[titles.iloc[tup[0]], tup[1]*100] for tup in sim_scores]

  return pd.DataFrame(data=movie_sim, columns=['Recommended movies similar to '+title, 'Similarity Score'], index=movie_sim_idx)


In [ ]:
cbr_recommendor('Toy Story')

Recommended movies similar to Toy Story  Similarity Score
2502                             Toy Story 2         24.804050
7535                             Toy Story 3         23.080578
6193                  The 40 Year Old Virgin         10.628646
2547                         Man on the Moon         10.163067
6627                            Factory Girl          8.047870

In [ ]:
cbr_recommendor('The Dark Knight',15)

Recommended movies similar to The Dark Knight  Similarity Score
7931                         The Dark Knight Rises         17.183830
132                                 Batman Forever         12.238593
1113                                Batman Returns         10.085391
8227       Batman: The Dark Knight Returns, Part 2          8.463275
7565                    Batman: Under the Red Hood          8.413981
524                                         Batman          8.161776
7901                              Batman: Year One          7.781253
2579                  Batman: Mask of the Phantasm          6.961079
2696                                           JFK          6.155398
8165       Batman: The Dark Knight Returns, Part 1          6.094795
6144                                 Batman Begins          5.393485
7933            Sherlock Holmes: A Game of Shadows          5.391813
5511                               To End All Wars          5.306932
4489                                         Q & A          5.131479
7344                           Law Abiding Citizen          5.011333

### ***Conclusion:***

So, for ***The Dark Knight***, the system is able to identify it as a Batman film and subsequently recommend other Batman films as its top recommendations.

But unfortunately, that is all this system can do at the moment. This is not of much use to most user as it doesn't take into considerations very important features such as ***cast, crew, director and genre***, which determine the rating and the popularity of a movie. Someone who liked ***The Dark Knight*** probably likes it more because of Nolan and would hate Batman Forever and every other substandard movie in the Batman Franchise.

Therefore, we need more suggestive metadata than Overview and Tagline. 

Now, we will build more sophisticated recommender that takes genre, keywords, cast and crew into consideration.

# ***Metadata Based Recommmendor (MBR)***

To build a standard metadata based content recommender, we need to merge our current dataset with the crew and the keyword datasets. 

**Let us prepare this data as our first step.**

In [ ]:
credits= pd.read_csv(r'/content/drive/MyDrive/DATA SCIENCE/Netflix Analysis/credits.csv')
keywords= pd.read_csv(r'/content/drive/MyDrive/DATA SCIENCE/Netflix Analysis/keywords.csv')

In [ ]:
credits.head(1).T 
#cast: dict. of cast_id, character (in movie), credit_id, gender (2 for male, 1 for female), name (actor/ voice artist), order (index of dict.), profile path.
#crew: credit_id, department (Directing, Writing, Editing, Sound,... etc.), gender (2 for male, 1 for female), id (crew_id), job (job in/for movie like, director, 
#      screenplay, etc), name (crew member name), profile path.
#id: movie id

0
cast  [{'cast_id': 14, 'character': 'Woody (voice)',...
crew  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...
id                                                  862

In [ ]:
keywords.head(1).T
#movie based keywords.

0
id                                                      862
keywords  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...

In [ ]:
#Merging md, keywords and credits ON 'id'
md['id']= md.id.astype('int')
keywords['id']= keywords.id.astype('int')
credits['id']= credits.id.astype('int')

In [ ]:
md.shape

(45463, 23)

In [ ]:
#merging 
md= md.merge(credits, on= 'id')
md= md.merge(keywords, on= 'id')
md.shape

(46628, 26)

In [ ]:
smd = md[md.id.isin(links_sml)]
smd.shape

(9219, 26)

*Thus, 46628 reduced to 9219.* 

*Much more compute efficient*

In [ ]:
smd['year']=pd.to_datetime(smd.release_date, errors='coerce').apply(lambda x: str(x).split('-')[0])

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
smd.head(3)

id    imdb_id             title original_language  \
0    862  tt0114709         Toy Story                en   
1   8844  tt0113497           Jumanji                en   
2  15602  tt0113228  Grumpier Old Men                en   

                                    spoken_languages  \
0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2           [{'iso_639_1': 'en', 'name': 'English'}]   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   

                               homepage                        genres  \
0  http://toystory.disney.com/toy-story   [Animation, Comedy, Family]   
1                                   NaN  [Adventure, Fantasy, Family]   
2                                   NaN             [Romance, Comedy]   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   

                               belongs_to_collection  ...  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...   
1                                                NaN  ...   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...  ...   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   

                                production_countries  video  vote_average  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...  False           7.7   
1  [{'iso_3166_1': 'US', 'name': 'United States o...  False           6.9   
2  [{'iso_3166_1': 'US', 'name': 'United States o...  False           6.5   

   vote_count                                               cast  \
0      5415.0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1      2413.0  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2        92.0  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...   

                                            keywords  year  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...  1995  
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...  1995  
2  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...  1995  

[3 rows x 27 columns]

    Now in our data--------------------

**Crew:** 

From the crew, we will only pick the director as our feature since the others don't contribute that much to the feel of the movie.

**Cast:** 

Choosing Cast is a little more tricky. Lesser known actors and minor roles do not really affect people's opinion of a movie. Therefore, we must only select the major characters and their respective actors. Arbitrarily we will choose the top 3 actors that appear in the credits list.

In [ ]:
#extracting cast (top 3), crew (only director), and keywords 
smd['cast']= smd['cast'].apply(literal_eval).apply(lambda x: [cast['name'] for cast in x[:3]])
smd['crew']= smd['crew'].apply(literal_eval).apply(lambda x: [crew['name'] for crew in x if crew['job']=='Director'])
smd['crew']=smd['crew'].apply(lambda x: x*4) # NOTE: twiking a bit to give high priority to crew/directors

smd['keywords']= smd['keywords'].apply(literal_eval).apply(lambda x: [key['name'] for key in x])

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

In [ ]:
smd.head()

id    imdb_id                        title original_language  \
0    862  tt0114709                    Toy Story                en   
1   8844  tt0113497                      Jumanji                en   
2  15602  tt0113228             Grumpier Old Men                en   
3  31357  tt0114885            Waiting to Exhale                en   
4  11862  tt0113041  Father of the Bride Part II                en   

                                    spoken_languages  \
0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2           [{'iso_639_1': 'en', 'name': 'English'}]   
3           [{'iso_639_1': 'en', 'name': 'English'}]   
4           [{'iso_639_1': 'en', 'name': 'English'}]   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
3  Friends are the people who let you be yourself...   
4  Just When His World Is Back To Normal... He's ...   

                               homepage                        genres  \
0  http://toystory.disney.com/toy-story   [Animation, Comedy, Family]   
1                                   NaN  [Adventure, Fantasy, Family]   
2                                   NaN             [Romance, Comedy]   
3                                   NaN      [Comedy, Drama, Romance]   
4                                   NaN                      [Comedy]   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                               belongs_to_collection  ...  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...   
1                                                NaN  ...   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...  ...   
3                                                NaN  ...   
4  {'id': 96871, 'name': 'Father of the Bride Col...  ...   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg   
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries  video  vote_average  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...  False           7.7   
1  [{'iso_3166_1': 'US', 'name': 'United States o...  False           6.9   
2  [{'iso_3166_1': 'US', 'name': 'United States o...  False           6.5   
3  [{'iso_3166_1': 'US', 'name': 'United States o...  False           6.1   
4  [{'iso_3166_1': 'US', 'name': 'United States o...  False           5.7   

   vote_count                                               cast  \
0      5415.0                [Tom Hanks, Tim Allen, Don Rickles]   
1      2413.0     [Robin Williams, Jonathan Hyde, Kirsten Dunst]   
2        92.0         [Walter Matthau, Jack Lemmon, Ann-Margret]   
3        34.0  [Whitney Houston, Angela Bassett, Loretta Devine]   
4       173.0         [Steve Martin, Diane Keaton, Martin Short]   

                                                crew  \
0  [John Lasseter, John Lasseter, John Lasseter, ...   
1  [Joe Johnston, Joe Johnston, Joe Johnston, Joe...   
2  [Howard Deutch, Howard Deutch, Howard Deutch, ...   
3  [Forest Whitaker, Forest Whitaker, Forest Whit...   
4  [Charles Shyer, Charles Shyer, Charles Sh

In [ ]:
#removing white spaces from cast and crew
smd['cast']=smd['cast'].apply(lambda x: [cast.replace(' ','') for cast in x])
smd['crew']=smd['crew'].apply(lambda x: [crew.replace(' ','') for crew in x])

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
#keywords
#converting the following to vertical(stack) columns with index as keywords and droping the level 1 i.e column index
s = smd.apply(lambda x: pd.Series(x['keywords'], index=x['keywords']), axis=1).stack().reset_index(level=1, drop=True) 
s.name='kywrd' 

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
s=s.value_counts() #counting frequency 
s[:10] 

independent film        610
woman director          550
murder                  399
duringcreditsstinger    327
based on novel          318
violence                264
love                    222
sex                     219
musical                 219
suspense                212
Name: kywrd, dtype: int64

***Keywords*** occur in frequencies ranging from 1 to 610. We do not have any use for keywords that occur only once. Therefore, these can be safely removed. 

Finally, we will convert every word to its stem so that words such as ***Dogs and Dog are considered the same***.

In [ ]:
#Removed word occurrence less than 2 
s=s[s>1] 

In [ ]:
#converting every word to its stem 
from nltk.stem.snowball import SnowballStemmer
stemmer= SnowballStemmer('english')
stemmer.stem('dogs') #example

'dog'

In [ ]:
#coverting same keywords to single stem
smd['keywords']=smd['keywords'].apply(lambda x: [stemmer.stem(word) for word in x if word in s])
smd['keywords']= smd['keywords'].apply(lambda x: [word.replace(' ','').lower() for word in x])
smd['keywords']

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


0        [jealousi, toy, boy, friendship, friend, rival...
1        [boardgam, disappear, basedonchildren'sbook, n...
2                   [fish, bestfriend, duringcreditssting]
3        [basedonnovel, interracialrelationship, single...
4        [babi, midlifecrisi, confid, age, daughter, mo...
                               ...                        
40952                                         [friendship]
41172                                          [bollywood]
41225                                          [bollywood]
41391     [monster, godzilla, giantmonst, destruct, kaiju]
41669                                 [music, documentari]
Name: keywords, Length: 9219, dtype: object

In [ ]:
smd['feature_classification']= smd.crew + smd.cast + smd.keywords + smd.genres 
smd['feature_classification']= smd.feature_classification.apply(lambda x: ' '.join(x))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:
smd.feature_classification

0        JohnLasseter JohnLasseter JohnLasseter JohnLas...
1        JoeJohnston JoeJohnston JoeJohnston JoeJohnsto...
2        HowardDeutch HowardDeutch HowardDeutch HowardD...
3        ForestWhitaker ForestWhitaker ForestWhitaker F...
4        CharlesShyer CharlesShyer CharlesShyer Charles...
                               ...                        
40952    GreggChampion GreggChampion GreggChampion Greg...
41172    TinuSureshDesai TinuSureshDesai TinuSureshDesa...
41225    AshutoshGowariker AshutoshGowariker AshutoshGo...
41391    HideakiAnno ShinjiHiguchi HideakiAnno ShinjiHi...
41669    RonHoward RonHoward RonHoward RonHoward PaulMc...
Name: feature_classification, Length: 9219, dtype: object

*Now, we will vectorize and with the help of cosine similarity find recommendation movies based on* ***cast, crew, keywords and genres***

In [ ]:
#vectorize the feature classification
vectors2= tfidfvectors.fit_transform(smd.feature_classification)

In [ ]:
cos_sim2= cosine_similarity(vectors2, vectors2)
cos_sim_df= pd.DataFrame(data= cos_sim2)
cos_sim_df.head()

0         1         2         3         4         5         6     \
0  1.000000  0.003309  0.001320  0.001265  0.001145  0.000000  0.001328   
1  0.003309  1.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2  0.001320  0.000000  1.000000  0.003956  0.001332  0.000000  0.004153   
3  0.001265  0.000000  0.003956  1.000000  0.001277  0.000719  0.003981   
4  0.001145  0.000000  0.001332  0.001277  1.000000  0.000000  0.001340   

       7         8         9     ...  9209      9210      9211     9212  \
0  0.003841  0.000000  0.000000  ...   0.0  0.000000  0.004426  0.00000   
1  0.007277  0.003052  0.002418  ...   0.0  0.000000  0.006209  0.00000   
2  0.000000  0.000000  0.000000  ...   0.0  0.000000  0.000000  0.00000   
3  0.001065  0.000000  0.000000  ...   0.0  0.000792  0.000864  0.00068   
4  0.000000  0.000000  0.000000  ...   0.0  0.000000  0.000000  0.00000   

       9213      9214      9215      9216      9217  9218  
0  0.000884  0.007334  0.000000  0.000000  0.000000   0.0  
1  0.000000  0.000000  0.000000  0.002957  0.002043   0.0  
2  0.001028  0.000000  0.002322  0.002434  0.000000   0.0  
3  0.000986  0.000975  0.002226  0.003295  0.000664   0.0  
4  0.000892  0.000000  0.000000  0.000000  0.000000   0.0  

[5 rows x 9219 columns]

In [ ]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title']) #find index from title
smd

index      id    imdb_id  \
0         0     862  tt0114709   
1         1    8844  tt0113497   
2         2   15602  tt0113228   
3         3   31357  tt0114885   
4         4   11862  tt0113041   
...     ...     ...        ...   
9214  40952  159550  tt0255313   
9215  41172  392572  tt5165344   
9216  41225  402672  tt3859980   
9217  41391  315011  tt4262980   
9218  41669  391698  tt2531318   

                                                  title original_language  \
0                                             Toy Story                en   
1                                               Jumanji                en   
2                                      Grumpier Old Men                en   
3                                     Waiting to Exhale                en   
4                           Father of the Bride Part II                en   
...                                                 ...               ...   
9214                     The Last Brickmaker in America                en   
9215                                             Rustom                hi   
9216                                       Mohenjo Daro                hi   
9217                                      Shin Godzilla                ja   
9218  The Beatles: Eight Days a Week - The Touring Y...                en   

                                       spoken_languages  \
0              [{'iso_639_1': 'en', 'name': 'English'}]   
1     [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2              [{'iso_639_1': 'en', 'name': 'English'}]   
3              [{'iso_639_1': 'en', 'name': 'English'}]   
4              [{'iso_639_1': 'en', 'name': 'English'}]   
...                                                 ...   
9214                                                 []   
9215            [{'iso_639_1': 'hi', 'name': 'हिन्दी'}]   
9216            [{'iso_639_1': 'hi', 'name': 'हिन्दी'}]   
9217  [{'iso_639_1': 'it', 'name': 'Italiano'}, {'is...   
9218           [{'iso_639_1': 'en', 'name': 'English'}]   

                                                tagline  \
0                                                   NaN   
1             Roll the dice and unleash the excitement!   
2     Still Yelling. Still Fighting. Still Ready for...   
3     Friends are the people who let you be yourself...   
4     Just When His World Is Back To Normal... He's ...   
...                                                 ...   
9214                                                NaN   
9215  Decorated Officer. Devoted Family Man. Defendi...   
9216                                                NaN   
9217                    A god incarnate. A city doomed.   
9218            The band you know. The story you don't.   

                                   homepage  \
0      http://toystory.disney.com/toy-story   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   
...                                     ...   
9214                                    NaN   
9215                                    NaN   
9216                                    NaN   
9217                                    NaN   
9218  http://www.thebeatlesliveproject.com/   

                                                 genres  \
0                           [Animation, Comedy, Family]   
1                          [Adventure, Fantasy, Family]   
2                                     [Romance, Comedy]   
3                              [Comedy, Drama, Romance]   
4                                              [Comedy]   
...                                                 ...   
9214                                            [Drama]   
9215                                [Thriller, Romance]   
9216               [Adventure, Drama, History, Romance]   
9217  [Action, Adventure, Drama, Horror, Science Fic...   
9218                               [Docu

In [ ]:
indices['The Dark Knight']

6981

In [ ]:
cbr_recommendor('The Dark Knight',15) # NOTE: for cbr_rcommendor smd, indices, and cos_sim_df is required

Recommended movies similar to The Dark Knight  Similarity Score
8031                         The Dark Knight Rises         49.540651
6218                                 Batman Begins         48.034923
6623                                  The Prestige         41.748438
2085                                     Following         39.154891
7648                                     Inception         39.110289
4145                                      Insomnia         38.406407
3381                                       Memento         36.409264
8613                                  Interstellar         29.870084
1134                                Batman Returns         13.482333
7659                    Batman: Under the Red Hood         11.826378
1260                                Batman & Robin          7.282574
2131                                      Superman          7.183153
9024            Batman v Superman: Dawn of Justice          6.483246
524                                         Batman          6.378825
132                                 Batman Forever          6.018984

In [ ]:
cbr_recommendor('The Godfather') # NOTE: for cbr_rcommendor smd, indices, and cos_sim_df is required

Recommended movies similar to The Godfather  Similarity Score
3616               Tucker: The Man and His Dream         51.306872
1992                       Peggy Sue Got Married         47.077791
3705                             The Cotton Club         46.413371
4518                          One from the Heart         45.975554
5867                                 Rumble Fish         45.667854

#### ***Conclusion:***

Thus, the recommendations seems to have recognized other Christopher Nolan movies (due to the high weightage given to director) and put them as top recommendations. 

Thus, able to satisfy the users intrest to see fav. directors movies

#***Popularity and Ratings***

One thing to note that the recommendation system is that it recommends movies regardless of ratings and popularity. So, movies like ***Batman and Robin*** has a lot of similar characters as compared to ***The Dark Knight*** but it was a terrible movie that shouldn't be recommended to anyone.

Therefore, we need a mechanism to remove bad movies and return movies which are popular and have had a good critical response.

So, based on similarity scores and calculate the vote of the 95th percentile movie. Then, using the value of  m to calculate the weighted rating of each movie using IMDB's formula like we did in the Simple Recommender section fr top movies.

In [ ]:
#improve in movie recommendor (cbr_recommendor) to new and better recommmendor
#i.e., including cast, crew, keywords and genres --> Movie Description based recommnedor

def movie_recommendor(title, top_n =10):
  idx= indices[title]

  sim_scores= list(enumerate(cos_sim_df[idx]))
  sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
  sim_scores = sim_scores[1:26] #removing the first one

  movie_sim_idx=[tup[0] for tup in sim_scores]

  movies_sim= smd.iloc[movie_sim_idx][['title', 'crew','cast', 'genres', 'year','vote_count', 'vote_average']]
  movies_sim['crew']= smd['crew'].apply(lambda x: list(set(x)))
  movies_sim['vote_count']= movies_sim[movies_sim.vote_count.notnull()]['vote_count'].astype('int')
  movies_sim['vote_average']= movies_sim[movies_sim.vote_average.notnull()]['vote_average'].astype('int')


  m= movies_sim.vote_count.quantile(0.60)
  C= movies_sim.vote_average.mean()

  qualified_movies= movies_sim[(movies_sim.vote_count>=m) & (movies_sim.vote_count.notnull()) & (movies_sim.vote_average.notnull())]

  v= qualified_movies.vote_count
  R= qualified_movies.vote_average
  wr= ((v/(v+m)) * R) + (m/(v+m) * C)

  qualified_movies['wr'] = wr
  return qualified_movies.sort_values('wr', ascending=False).head(top_n)


In [ ]:
movie_recommendor('The Dark Knight')

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:28: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


title                crew  \
7648                           Inception  [ChristopherNolan]   
8613                        Interstellar  [ChristopherNolan]   
6623                        The Prestige  [ChristopherNolan]   
3381                             Memento  [ChristopherNolan]   
8031               The Dark Knight Rises  [ChristopherNolan]   
6218                       Batman Begins  [ChristopherNolan]   
7583                            Kick-Ass     [MatthewVaughn]   
8467                          Kick-Ass 2        [JeffWadlow]   
8419                        Man of Steel        [ZackSnyder]   
9024  Batman v Superman: Dawn of Justice        [ZackSnyder]   

                                                   cast  \
7648  [LeonardoDiCaprio, JosephGordon-Levitt, EllenP...   
8613  [MatthewMcConaughey, JessicaChastain, AnneHath...   
6623         [HughJackman, ChristianBale, MichaelCaine]   
3381        [GuyPearce, Carrie-AnneMoss, JoePantoliano]   
8031          [ChristianBale, MichaelCaine, GaryOldman]   
6218          [ChristianBale, MichaelCaine, LiamNeeson]   
7583  [AaronTaylor-Johnson, ChloëGraceMoretz, Christ...   
8467  [AaronTaylor-Johnson, ChloëGraceMoretz, Christ...   
8419            [HenryCavill, AmyAdams, MichaelShannon]   
9024                [BenAffleck, HenryCavill, GalGadot]   

                                                 genres  year  vote_count  \
7648  [Action, Thriller, Science Fiction, Mystery, A...  2010       14075   
8613                [Adventure, Drama, Science Fiction]  2014       11187   
6623                         [Drama, Mystery, Thriller]  2006        4510   
3381                                [Mystery, Thriller]  2000        4168   
8031                   [Action, Crime, Drama, Thriller]  2012        9263   
6218                             [Action, Crime, Drama]  2005        7511   
7583                                    [Action, Crime]  2010        4747   
8467                         [Action, Adventure, Crime]  2013        2275   
8419      [Action, Adventure, Fantasy, Science Fiction]  2013        6462   
9024                       [Action, Adventure, Fantasy]  2016        7189   

      vote_average        wr  
7648             8  7.789373  
8613             8  7.743924  
6623             8  7.488994  
3381             8  7.461537  
8031             7  6.892642  
6218             7  6.873267  
7583             7  6.822823  
8467             6  6.216163  
8419             6  6.111639  
9024             5  5.337064

#### ***NOTE:***

Things that can improve the recommendor:

*   Range of sim_scores to be consider
*   m i.e., minimum value for vote count to be consider
*   keywords: increase the frequency of occurrence (here, more than 1 is taken)






#### ***Conclusion:***

The Recommendor on movies ranges till top 26 with 65th percentile in vote_count and contain similar keywords more than once is giving user desired result of movies directed by ***Christopher Nolan***.

# ***Collaborative Filtering (CBF)***

Our content based engine suffers from some severe limitations. It is only capable of suggesting movies which are close to a certain movie. That is, it is not capable of capturing tastes and providing recommendations across genres.

Also, the engine that we built is not really personal in that it doesn't capture the personal tastes and biases of a user. Anyone querying our engine for recommendations based on a movie will receive the same recommendations for that movie, regardless of who s/he is.

Therefore, in this section, we will use a technique called **Collaborative Filtering** to make recommendations to Movie Watchers. Collaborative Filtering is based on the idea that users similar to a me can be used to predict how much I will like a particular product or service those users have used/experienced but I have not.

Now, we will use Singular Value Decomposition (SVD) and K-Nearest Neighbours (KNN) which are very powerful Algorithms present in Surpise lib as well as scikitLearn (not use bcz of no predict fuction) to minimise RMSE (Root Mean Square Error) and give great recommendations.

In [ ]:
#read ratings for users from ratings_small as contain less records which is benifitial for computation
us_ratings= pd.read_csv(r'/content/drive/MyDrive/DATA SCIENCE/Netflix Analysis/ratings_small.csv')
us_ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [ ]:
us_ratings.shape

(100004, 4)

In [ ]:
! pip install numpy
! pip install scikit-surprise

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
#Simple Python RecommendatIon System Engine (SURPRISE)
from surprise import Reader, SVD, KNNWithMeans, Dataset

data= Dataset.load_from_df(us_ratings[['userId', 'movieId', 'rating']], Reader())

In [ ]:
'''
SVD
    n_epochs – The number of iteration of the SGD procedure. Default is 20.
    lr_all – The learning rate for all parameters. Default is 0.005.
    reg_all – The regularization term for all parameters. Default is 0.02.

KNN

'''
from surprise.model_selection import cross_validate

algo=SVD(random_state=40)
cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8933  0.8988  0.8991  0.8947  0.8967  0.8965  0.0022  
MAE (testset)     0.6897  0.6924  0.6905  0.6906  0.6907  0.6908  0.0009  
Fit time          5.71    8.27    10.02   5.14    5.15    6.86    1.96    
Test time         0.19    0.38    0.40    0.15    0.15    0.25    0.11    


{'test_rmse': array([0.89326645, 0.89876966, 0.89906151, 0.89474337, 0.89669279]),
 'test_mae': array([0.68967834, 0.69240132, 0.69052022, 0.69057448, 0.69070569]),
 'fit_time': (5.706661701202393,
  8.273340463638306,
  10.015141248703003,
  5.141899108886719,
  5.150862216949463),
 'test_time': (0.1907367706298828,
  0.3817713260650635,
  0.40161705017089844,
  0.14585113525390625,
  0.14843511581420898)}

We get a mean Root Mean Sqaure Error of 0.8968 which is more than good enough for our case. Let us now train on our dataset and arrive at predictions.

In [ ]:
trainset = data.build_full_trainset()
algo.fit(trainset)

In [ ]:
us_ratings[us_ratings['movieId'] == 300]

userId  movieId  rating   timestamp
50          2      300     3.0   835355532
1038       15      300     4.0  1054449869
3193       19      300     3.0   855193220
4043       23      300     4.0  1149868544
5097       30      300     5.0   945122065
...       ...      ...     ...         ...
96205     641      300     4.0   834636572
96610     647      300     5.0   947292218
98451     659      300     5.0   834598140
98679     662      300     3.0   839022324
99762     667      300     4.0   847271534

[90 rows x 4 columns]

In [ ]:
algo.predict(uid=1, iid= 302)

Prediction(uid=1, iid=302, r_ui=None, est=2.9865519836420185, details={'was_impossible': False})

***`Conclusion:`***

For movie with ID 302, we get an estimated prediction of **2.759**. (This will affect due to random_state in algo, uid i.e., userId, iid i.e., itemId (here, movieId)

One startling feature of this recommender system is that it doesn't care what the movie is (or what it contains). 

It works purely on the basis of an assigned movie ID and tries to predict ratings based on how the other users have predicted the movie.

# ***Hybrid Recommender***

It is the combination of both Content Based Filtering and as well as Collaberative based Filtering.

***`This is how it will work:`***

***Input:*** User ID and the Title of a Movie

***Output:*** Similar movies sorted on the basis of expected ratings by that particular user.

In [ ]:
# importing basic packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from ast import literal_eval